utils/misc/rl_utils/rl_utils/utils/observation_collector/observation_units/pedsim_collector_unit.py

```diff
@@ -0,0 +1,138 @@
+        _agent_types (List[str]): A list of agent types.
+    """
+
+    _agent_states: List[AgentState]
+    _agent_states_sub: rospy.Subscriber
+
+    _agent_poses: List[Pose]
+    _agent_twists: List[Twist]
+    _agent_types: List[str]
+
+    def __init__(
+        self, ns: Namespace, observation_manager: "ObservationCollector"
+    ) -> None:
+        """
+        Initializes the PedsimStateCollectorUnit.
+
+        Args:
+            ns (Namespace): The namespace for the simulation.
+            observation_manager (ObservationCollector): The observation manager containing the unit.
+        """
+        super().__init__(ns.simulation_ns, observation_manager)
+        self._agent_states = AgentStates()
+        self._agent_states_sub: rospy.Subscriber = None
+
+        self._agent_poses = []
+        self._agent_twists = []
+        self._agent_types = []
+
+    def init_subs(self):
+        """
+        Initializes the subscriber for agent states.
+        """
+        self._agent_states_sub = rospy.Subscriber(
+            self._ns(TOPICS.PEDSIM_STATES), AgentStates, self._cb_pedsim_states
+        )
+
+    def get_observations(
+        self, obs_dict: Dict[str, Any], *args, **kwargs
+    ) -> Dict[str, Any]:
+        """
+        Retrieves the observations and updates the observation dictionary.
+
+        Args:
+            obs_dict (Dict[str, Any]): The observation dictionary.
+
+        Returns:
+            Dict[str, Any]: The updated observation dictionary.
+        """
+
+        obs_dict = super().get_observations(obs_dict, *args, **kwargs)
+        obs_dict.update(
+            {
+                OBS_DICT_KEYS.PEDSIM_POSES: self._agent_poses,
+                OBS_DICT_KEYS.PEDSIM_TWISTS: self._agent_twists,
+                OBS_DICT_KEYS.PEDSIM_TYPES: self._agent_types,
+            }
+        )
+        return obs_dict
+
+    def _cb_pedsim_states(self, pedsim_states_msg: AgentStates) -> None:
+        """
+        Callback function for processing received pedestrian simulation states.
+
+        Args:
+            pedsim_states_msg (AgentStates): The received pedestrian simulation states message.
+        """
+        self._agent_states = pedsim_states_msg.agent_states
+        self._agent_poses = PedsimStateCollectorUnit._get_agent_positions(
+            self._agent_states
+        )
+        self._agent_twists = PedsimStateCollectorUnit._get_agent_twists(
+            self._agent_states
+        )
+        self._agent_types = PedsimStateCollectorUnit._get_agent_types(
+            self._agent_states
+        )
+
+    @staticmethod
+    def _get_agent_positions(agent_states: List[AgentState]) -> List[Pose]:
+        """
+        Static method to extract agent positions from a list of agent states.
+
+        Args:
+            agent_states (List[AgentState]): A list of agent states.
+
+        Returns:
+            List[Pose]: A list of agent poses.
+        """
+        return [agent.pose for agent in agent_states]
```

Comment:
Keep in mind that `agent.pose.orientation` is (currently) bogus data. The correct yaw can be calculated from the velocity vector (which is also consistent with Gazebo) when needed.